esp-hal/src/clock/mod.rs

```diff
@@ -591,65 +599,135 @@ impl Clocks {
+            clocks_ll::enable_phy(true);
+        }
+
+        phy_clock_ref_counter.set(phy_clock_ref_count + 1);
+    });
 }
+#[cfg(any(bt, ieee802154, wifi))]
+fn decrease_phy_clock_ref_count_internal() {
+    critical_section::with(|cs| {
+        let phy_clock_ref_counter = PHY_CLOCK_REF_COUNTERS.borrow(cs);
 
+        let new_phy_clock_ref_count = phy_clock_ref_counter.get().checked_sub(1).expect("PHY clock ref count underflowed. Either you forgot a PhyClockGuard, or used ModemClockController::decrease_phy_clock_ref_count incorrectly.");
+
+        if new_phy_clock_ref_count == 0 {
+            clocks_ll::enable_phy(false);
+        }
+        phy_clock_ref_counter.set(new_phy_clock_ref_count);
+    });
+}
+/// Do any common initial initialization needed for the radio clocks
+#[inline]
+pub fn init_radio_clocks() {
+    clocks_ll::init_clocks();
+}
+#[instability::unstable]
+#[cfg(any(bt, ieee802154, wifi))]
+#[derive(Debug)]
+/// Prevents the PHY clock from being disabled.
+///
+/// As long as at least one [PhyClockGuard] exists, the PHY clock will remain
+/// active. To release this guard, you can either let it go out of scope or use
+/// [PhyClockGuard::release] to explicitly release it.
+pub struct PhyClockGuard<'d> {
+    _phantom: PhantomData<&'d ()>,
+}
 #[cfg(any(bt, ieee802154, wifi))]
-impl<'d> RadioClockController<'d> {
-    /// Create a new instance of the radio clock controller
+impl PhyClockGuard<'_> {
     #[instability::unstable]
-    pub fn new(rcc: crate::peripherals::RADIO_CLK<'d>) -> Self {
-        Self { _rcc: rcc }
+    #[inline]
+    /// Release the clock guard.
+    ///
+    /// The PHY clock will be disabled, if this is the last clock guard.
+    pub fn release(self) {}
+}
+#[cfg(any(bt, ieee802154, wifi))]
+impl Drop for PhyClockGuard<'_> {
+    fn drop(&mut self) {
+        decrease_phy_clock_ref_count_internal();
     }
+}
+#[cfg(any(bt, ieee802154, wifi))]
+#[instability::unstable]
+/// This trait provides common functionality for all
+pub trait ModemClockController<'d>: Sealed + 'd {
+    /// Enable the modem clock for this controller.
+    fn enable_modem_clock(&mut self, enable: bool);
 
-    /// Enable the PHY clocks
-    #[instability::unstable]
-    #[cfg(phy)]
-    #[inline]
-    pub fn enable_phy(&mut self, enable: bool) {
-        clocks_ll::enable_phy(enable);
+    /// Enable the PHY clock and acquire a [PhyClockGuard].
+    ///
+    /// The PHY clock will only be disabled, once all [PhyClockGuard]'s of all
+    /// modems were dropped.
+    fn enable_phy_clock(&self) -> PhyClockGuard<'d> {
+        increase_phy_clock_ref_count_internal();
+        PhyClockGuard {
+            _phantom: PhantomData,
+        }
     }
 
-    /// Enable the Bluetooth clocks
-    #[instability::unstable]
-    #[cfg(bt)]
-    #[inline]
-    pub fn enable_bt(&mut self, enable: bool) {
-        clocks_ll::enable_bt(enable);
+    /// Unsafely decrease the PHY clock reference count for this modem.
+    ///
+    /// # Safety
```

Comment:
This function is not `unsafe`, so you don't need a `# Safety` comment. On the other hand, if this may panic, you need a `# Panics`. We also shouldn't really recomment using mem::forget here, as we explicitly say not to use it in the crate docs.

Reply:
please don't, let's not encourage bad behaviour